### DISCUSSION TITLE
nestjs-docker-production

### DISCUSSION THREAD

**Discussion by giscus** [tomwray13/tomray.dev#19] [Blog Comments]
# nestjs-docker-production

Learn how to write a Dockerfile that creates a production optimized image using the NodeJS Alpine image and multistage builds.

https://www.tomray.dev/nestjs-docker-production#comments-container

**Comment by johnta0** [comment]
Thank you for writing this article.
It would be more helpful if there was `package.json` example provided.

Actually, I'm stuck getting the error: nest command not found.

**Reply by sinhaaritro** [reply]
My guess is that in the local environment you have nestcli globally installed. The node image doesn't have nestcli installed. 

If this is the problem then I can think of 2 solution:
1. Install @nestjs/cli globally in the node image.
2. Add @nestjs/cli as devDependencies in package.json

**Reply by tomwray13** [reply]
@johnta0 Hey there! I'm glad you liked this article.

[This repo](https://github.com/tomwray13/nest-cloud-run) contains the same Dockerfile used in this tutorial.

I had a similar error when I was using an older node base image. Do you still get the error when using `node:18` as the base image?

**Reply by Sarps** [reply]
This has nothing to do with the node version. Just like @sinhaaritro said, nest cli is in devDependency but your dockerfile presumably ran install with the production flag so the cli wasn't installed. 

And that is, of course, the right thing to do. You don't need the cli in production. 
So I would not advise you to take the flag off, neither would I ask you to move the dependency up.

This is my solution.

    1. Look through your package.json, you'll find that nest has made provision for a production start script which doesn't depend on the cli. It's called "start:prod". 
    2. Change the last part of your dockerfile from "npm run start" to "npm run start:prod".

**Comment by jeffersonest** [comment]
How to use Https?

**Comment by okerx** [comment]
IT IS THE ULTIMATE GUIDE!
Saved me tons of time, Thanks!

Here's a slightly modified version of Dockerfile that works well with Yarn,

```Dockerfile
###################
# BUILD FOR LOCAL DEVELOPMENT
###################
FROM node:18-alpine As development

WORKDIR /usr/src/app

COPY --chown=node:node package.json ./
COPY --chown=node:node yarn.lock ./

RUN yarn

COPY --chown=node:node . .

USER node


###################
# BUILD FOR PRODUCTION
###################
FROM node:18-alpine As build

WORKDIR /usr/src/app

COPY --chown=node:node package.json ./
COPY --chown=node:node yarn.lock ./

COPY --chown=node:node --from=development /usr/src/app/node_modules ./node_modules

COPY --chown=node:node . .

RUN yarn build

ENV NODE_ENV production

RUN yarn install --frozen-lockfile --production && yarn cache clean

USER node


###################
# PRODUCTION
###################
FROM node:18-alpine As production

COPY --chown=node:node --from=build /usr/src/app/node_modules ./node_modules
COPY --chown=node:node --from=build /usr/src/app/dist ./dist

CMD [ "node", "dist/main.js" ]
```

**Reply by tomwray13** [reply]
Thank you @ammar-oker! I'll try to update the blog post some point soon to have a `yarn` option.

**Comment by santiagazo** [comment]
Solid article. Loved how clear and simple it was. Slight typo in the word `producton` in the part about env variables.

**Comment by santiagazo** [comment]
I just found your other article for NestJS + PostgreSQL! Exactly what I was looking for! Thanks a ton! I read this one and it helped a lot, saved me lots of time, but then I needed Postgres too. Lo and behold same author and same quality. Nice.
https://www.tomray.dev/nestjs-docker-compose-postgres

**Reply by tomwray13** [reply]
So glad you liked it! I'm going to add some content around Redis to the docker compose post soon, too 😃

**Comment by bbangjooo** [comment]
Absolutely best guide I've ever seen. I'll keep following you!

**Reply by tomwray13** [reply]
Thank you @bbangjooo!

**Comment by tomasz-jankowski** [comment]
I'm getting following error:

```
npm ERR! code ENOENT
npm ERR! syscall open
npm ERR! path //package.json
npm ERR! errno -2
npm ERR! enoent ENOENT: no such file or directory, open '//package.json'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent 

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2022-08-16T20_57_01_139Z-debug-0.log
```

**Reply by NelsonCruzG** [reply]
In case you're using docker-compose, make sure your app process service context is pointing to the root folder of your project or wherever your package.json file is placed. Hope you can solve it.

**Comment by thanhn1012** [comment]
Thank you so much!

**Comment by Sarps** [comment]
A few comments:

1. I wouldn't advised copying your node modules from build stage into production. Re-run npm install with production flag to prune unnecessary dependencies which may otherwise affect final image size.

2. The start command, "node dist/main.js". You should rather create an npm script, put the command in there and start from the npm script instead (for all nodejs projects). In the case of nestjs, there's already a start script in package.json to run your application. Leverage on that.  It's cleaner and safer, and it doesn't couple tightly your Dockerfile with your directory structure. Which is to say, if nest decides to change, the build directory from "dist" to "build", your Dockerfile isn't affectted.

**Reply by flarno11** [reply]
with an npm script the `package.json` file is needed in the production stage which I assume is omitted intentionally.

**Comment by ak99372** [comment]
If we're talking about prod optimization why do the build and install packages inside docker? Wouldn't it better to bundle/minify all into single index.js first and then for docker only copy & run? It would be better even for start time.

**Reply by Sarps** [reply]
I don't think that was the goal of this guide. Bundling everything into a single index.js would require modifying the nestjs build configuration and is not something to meddle with if you're not very well-versed in the bundler ecosystem. His goal for writing this guide was to help someone who has already started a nestjs project (or is yet to) prepare a Dockerfile that is properly oriented for production container deployment. So, it would be nice if you can align all feedbacks in that direction.

**Comment by pazoozooCH** [comment]
Is there any compelling reason to change the owner/group of the app files to node:node? Why not just leave them at root:root with read-permissions for everybody? Node will still be able to run the app if it can just *read* the files - no need for write access. And if we should require write access to the FS for some reason, we'll most likely have a volume for this anyway...

**Reply by Sarps** [reply]
There is no "compelling" reason to run as "node:node" but for security reasons, it is much advised to run all applications within a container as non-root (not just node applications) at all times, unless strictly impossible.

> Remember that a process running in a container is no different from other process running on Linux, except it has a small piece of metadata that declares that it’s in a container.
https://medium.com/@mccode/processes-in-containers-should-not-run-as-root-2feae3f0df3b

**Reply by pazoozooCH** [reply]
I'm not talking about the running process. I'm talking about the JS files. I'm not questioning the running user, that's perfectly fine. I'm questioning the changing of the user/owner of the app files. We can still run the application (as non-root user *node*) if they belong to root:root and are only readable (and not writable) by the node user...

Or in other words: you can safely remove all the `--chown` parts in the `COPY` sections and the container will still run just fine (as user `node`).

**Reply by flarno11** [reply]
for security reasons no `--chown` should be used and the files should be kept as `root` to avoid any in-place modifications/tampering

**Comment by ilclaudio** [comment]
Hi,
thank you for your guide.
What about creating a single executable to minimize the ram used by node?
Is it possible according to you?
Thank you
cld

**Comment by igorlacerdak** [comment]
Hello Tom, how are you?

I'm Brazilian and a developer in a store company,
I saw your article on uploading an application in performatic docker using certain instructions in the Dockerfile.
And congratulations for the content, the article was excellent.

I have a doubt, which I could not research properly and was unsuccessful.

Here's the thing, my situation, I have an isolated container that runs the service (REDIS) with that I try to create another container for my API accessing redis but the API can't connect/communicate with the redis container.

All forums inform to use docker-compose, but if I have the container I should be able to access the local redis normally.
Do you have a suggestion on how I can do this communication? I don't have the need to create redis in compose, because every app would create a redis, it would be unnecessary several redis images.

**Comment by neeraj-Brightly** [comment]
hi Tom, 
I am facing some issue in running production image. I have tried removing node_modules and a fresh ci, but it still occurs. 
Here is the error

``` 
[Nest] 1  - 01/06/2023, 6:01:46 AM   ERROR [PackageLoader] No driver (HTTP) has been selected. In order to take advantage of the default driver, please, ensure to install the "@nestjs/platform-express" package ($ npm install @nestjs/platform-express).
```
In my package.json, Platform-express is a dependency and not a devDependency.

**Reply by neeraj-Brightly** [reply]
I was able to resolve it by doing 
```
RUN npm install --omit=dev && npm cache clean --force
```
Instead of 
```
RUN npm ci --only=production && npm cache clean --force
```
But I also tried doing

```
RUN npm ci --omit=dev && npm cache clean --force
```
which didn't work. 
I don't know what might be happening here.

**Comment by flarno11** [comment]
great guide!

but I think in the production stage you missed the `USER node`. prod would still run as `root` without it.

also `WORKDIR /usr/src/app` may be needed in the production stage, unless you really want to copy the files to `/`

**Comment by ivose** [comment]
If i use the last sample in a median-project. Then i receive error:

[+] Building 147.8s (12/15)
 => [internal] load build definition from Dockerfile                                         0.0s
 => => transferring dockerfile: 948B                                                         0.0s
 => [internal] load .dockerignore                                                            0.1s
 => => transferring context: 102B                                                            0.0s
 => [internal] load metadata for docker.io/library/node:18-alpine                            4.7s
 => CACHED [development 1/5] FROM docker.io/library/node:18-alpine@sha256:bc329c7332cffc30c  0.0s
 => [internal] load build context                                                            0.2s
 => => transferring context: 369.59kB                                                        0.1s
 => CACHED [development 2/5] WORKDIR /usr/src/app                                            0.0s
 => [development 3/5] COPY --chown=node:node package*.json ./                                0.2s
 => [development 4/5] RUN npm ci                                                            80.1s
 => [development 5/5] COPY --chown=node:node . .                                             0.2s
 => [build 4/7] COPY --chown=node:node --from=development /usr/src/app/node_modules ./node  27.7s
 => [build 5/7] COPY --chown=node:node . .                                                   0.2s
 => ERROR [build 6/7] RUN npm run build                                                     18.1s
------
 > [build 6/7] RUN npm run build:
#12 4.347
#12 4.347 > median@0.0.1 build
#12 4.347 > nest build
#12 4.347
#12 17.83 src/articles/entities/article.entity.ts:3:10 - error TS2305: Module '"@prisma/client"' h
as no exported member 'Article'.
#12 17.83
#12 17.83 3 import { Article } from '@prisma/client';
#12 17.83            ~~~~~~~
#12 17.83
#12 17.83 Found 1 error(s).
#12 17.83
#12 17.94 npm notice
#12 17.94 npm notice New minor version of npm available! 9.3.1 -> 9.4.1
#12 17.94 npm notice Changelog: <https://github.com/npm/cli/releases/tag/v9.4.1>
#12 17.94 npm notice Run `npm install -g npm@9.4.1` to update!
#12 17.94 npm notice
------
executor failed running [/bin/sh -c npm run build]: exit code: 1


WWhat to do?

**Comment by ikbenignace** [comment]
I do get: error: Environment variable not found: DATABASE_URL.

**Comment by nolawnchairs** [comment]
This Dockerfile actually causes some CI pipelines to fail. For me, this failed in BitBucket Pipelines, due to the way they are run in virtual machines, non-root file permissions can get lost between these different execution contexts.

There is no need to run the first two stages under the "node" user - root is fine, since it's an intermediate, ephemeral build. Only in the final production build should this be set with the "node" user. Doing this, the builds all pass as expected.
